Firebase.h

```diff
@@ -25,49 +25,108 @@
 #include <WiFiClientSecure.h>
 #include <ESP8266HTTPClient.h>
 
-// FirebaseError represents a Firebase API error with a code and a
-// message.
-class FirebaseError {
+//TODO(edcoyne) split these into multiple files.
+
+// Result from call to Firebase backend. ALWAYS check isError() before
+// expecting any data.
+class FirebaseResult {
  public:
-  operator bool() const { return _code < 0; }
-  int code() const { return _code; }
-  const String& message() const { return _message; }
-  void reset() { set(0, ""); }
-  void set(int code, const String& message) {
-    _code = code;
-    _message = message;
+  FirebaseResult(int status);
+  FirebaseResult(int status, const String& response);
+  FirebaseResult(const FirebaseResult& result);
+
+  // True if there was an error completeing call.
+  bool isError() const;
+  String errorMessage() const;
+
+  // True if http status code is 200(OK).
+  bool isOk() const;
+  // Message sent back from Firebase backend.
+  const String& response() const;
+
+  int httpStatus() const {
+    return status_;
   }
+
  private:
-  int _code = 0;
-  String _message = "";
+  int status_;
+  String response_;
 };
 
-// Firebase is the connection to firebase.
+// Low level connection to Firebase backend, you probably want the
+// Firebase class below.
+class FirebaseConnection {
+ public:
+  FirebaseConnection(const String& host);
+  FirebaseConnection& auth(const String& auth);
+
+  const String& host() {
+    return host_;
+  }
+
+  HTTPClient& httpClient(){
+    return http_;
+  }
+
+  String makeURL(const String& path);
+
+  FirebaseResult sendRequest(const char* method, const String& path, const String& value);
+  FirebaseResult sendRequest(const char* method, const String& path);
+
+  FirebaseResult sendRequestGetBody(const char* method, const String& path);
```

Comment:
rather than having separate `sendRequest*` method, maybe we could make getting the `body` optional?

So you would do something like `fbase.post().body()` if you care about the body or `fbase.delete()` if you don't.